File: numbers/src/main/java/org/dicio/numbers/parser/lexer/MatchedToken.java
package org.dicio.numbers.parser.lexer;

import java.util.Set;

public class MatchedToken extends Token {
    private final Set<String> categories;
    private DurationToken durationTokenMatch = null;

    MatchedToken(final String value,
                 final String spacesFollowing,
                 final Set<String> categories) {
        super(value, spacesFollowing);
        this.categories = categories;
    }

    public void setDurationTokenMatch(final DurationToken durationTokenMatch) {
        this.durationTokenMatch = durationTokenMatch;
    }

    @Override
    public boolean hasCategory(final String category) {
        return categories.contains(category);
    }

    @Override
    public boolean isDurationToken() {
        return durationTokenMatch != null;
    }

    @Override
    public DurationToken asDurationToken() {
        if (durationTokenMatch == null) {
            return super.asDurationToken();
        } else {
            return durationTokenMatch;
        }
    }
}

File: numbers/src/main/java/org/dicio/numbers/parser/lexer/MatchedToken.kt
package org.dicio.numbers.parser.lexer

open class MatchedToken internal constructor(
    value: String,
    spacesFollowing: String,
    private val categories: Set<String>
) : Token(value, spacesFollowing) {
    private var durationTokenMatch: DurationToken? = null

    fun setDurationTokenMatch(durationTokenMatch: DurationToken?) {
        this.durationTokenMatch = durationTokenMatch
    }

    override fun hasCategory(category: String): Boolean {
        return categories.contains(category)
    }

    override val asDurationToken: DurationToken?
        get() = durationTokenMatch
}
